newrelic/provider_test.go

```diff
@@ -97,43 +100,64 @@ func testAccPreCheck(t *testing.T) {
 	if !testAccAPMEntityCreated {
-		// Clean up old instances of the applications
-		testAccApplicationsCleanup(t)
-
-		// Create the application
-		testAccCreateApplication(t)
-
-		// We need to give the entity search engine time to index the app so
-		// we try to get the entity, and retry if it fails for a certain amount
-		// of time
-		client := entities.New(config.Config{
-			PersonalAPIKey: testAccAPIKey,
-		})
-		params := entities.EntitySearchQueryBuilder{
-			Name:   testAccExpectedApplicationName,
-			Type:   "APPLICATION",
-			Domain: "APM",
-		}
+		testAccCreateEntity(t, testAccExpectedApplicationName)
+		testAccAPMEntityCreated = true
+	}
+}
 
-		retryErr := resource.RetryContext(context.Background(), 30*time.Second, func() *resource.RetryError {
-			entityResults, err := client.GetEntitySearchWithContext(context.Background(), entities.EntitySearchOptions{}, "", params, []entities.EntitySearchSortCriteria{})
-			if err != nil {
-				return resource.RetryableError(err)
-			}
+func testAccSingleQuotedPreCheck(t *testing.T) {
+	testAccPreCheckEnvVars(t)
 
-			if entityResults.Count != 1 {
-				return resource.RetryableError(fmt.Errorf("Entity not found, or found more than one"))
-			}
+	// Clean up old data partitions
+	//testAccLogDataPartitionsCleanup(t)
+
+	// Cleaning up the Parsing rules
+	//testAccLogParsingRulesCleanup(t)
+
+	// Create a test application for use in tests in the "entity" data source
+	// comprising an apostrophe "'".
+	if !testAccAPMSingleQuotedEntityCreated {
+		testAccCreateEntity(t, testAccExpectedSingleQuotedApplicationName)
+		testAccAPMSingleQuotedEntityCreated = true
+	}
+}
+
+func testAccCreateEntity(t *testing.T, name string) {
+	// Clean up old instances of the applications
+	testAccApplicationsCleanup(t)
+
+	// Create the application, with the given 'name' in the argument.
+	testAccCreateApplication(t, name)
 
-			return nil
-		})
+	// We need to give the entity search engine time to index the app so
+	// we try to get the entity, and retry if it fails for a certain amount
+	// of time
+	client := entities.New(config.Config{
+		PersonalAPIKey: testAccAPIKey,
+	})
+	params := entities.EntitySearchQueryBuilder{
+		Name:   escapeSingleQuote(name),
+		Type:   "APPLICATION",
+		Domain: "APM",
+	}
 
-		if retryErr != nil {
-			t.Fatalf("Unable to find application entity: %s", retryErr)
+	retryErr := resource.RetryContext(context.Background(), 30*time.Second, func() *resource.RetryError {
+		entityResults, err := client.GetEntitySearchWithContext(context.Background(), entities.EntitySearchOptions{}, "", params, []entities.EntitySearchSortCriteria{})
+		if err != nil {
+			return resource.RetryableError(err)
 		}
 
-		testAccAPMEntityCreated = true
+		if entityResults.Count != 1 {
+			return resource.RetryableError(fmt.Errorf("Entity not found, or found more than one"))
+		}
+
+		return nil
+	})
+
+	if retryErr != nil {
+		t.Fatalf("Unable to find application entity: %s", retryErr)
 	}
+
```

Comment:
A note for the reviewer - since the diff looks a bit messy here, here is a brief explanation of the changes made. 

In the older code, the function `testAccPreCheck` would deal with the entire procedure of creating an entity. However, this pre-check function cannot be used as the `PreCheck` attribute in the test case `TestAccNewRelicSingleQuotedEntityData_Basic`, since we need an entity to be created with an apostrophe in its name exclusively for this test case. To facilitate this, the procedure handling the creation of an entity has been shifted to a new function, `testAccCreateEntity`, which is now called by both, the older pre-check function `testAccPreCheck`, and the pre-check function added to create an entity with an apostrophe in its name, `testAccSingleQuotedPreCheck`.  `testAccSingleQuotedPrecheck` is then used as the `PreCheck` function in the test case added, `TestAccNewRelicSingleQuotedEntityData_Basic`.